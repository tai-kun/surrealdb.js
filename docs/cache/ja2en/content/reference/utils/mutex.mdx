---
title: mutex
---

import { Aside } from "@astrojs/starlight/components"

The `mutex` decorator ensures that only one of the decorated class methods is running at any given time. It achieves this by executing them asynchronously with a concurrency of 1.

<Aside type="caution">
  While there are no issues when using the built JavaScript package, using the source code (TypeScript) of this package directly requires Stage 3 decorator support.
</Aside>

### Import

```ts
import { mutex } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function mutex(
  target: (...args: any) => PromiseLike<any>,
  context: ClassMethodDecoratorContext,
): (this: any, ...args: any) => Promise<any>;
```

#### Parameters

##### `target`

The class method to be decorated.

##### `context`

The Stage 3 class method decorator context.

#### Return Value

An asynchronous function that limits the concurrency level to 1.

#### Example

The following example demonstrates the difference in the timing of asynchronous processing with and without the `mutex` decorator:

```ts
import { mutex } from "@tai-kun/surrealdb/utils";

class Runner {
  async runWithoutMutex(ms: number, value: string) {
    await sleep(ms);
    console.log(value);
  }

  @mutex
  async runWithMutex(ms: number, value: string) {
    await sleep(ms);
    console.log(value);
  }
}

const runner = new Runner();
// without mutex
await Promise.all([
  runner.runWithoutMutex(1000, "A"),
  runner.runWithoutMutex(500, "B"),
  runner.runWithoutMutex(0, "C"),
]);
// with mutex
await Promise.all([
  runner.runWithMutex(1000, "A"),
  runner.runWithMutex(500, "B"),
  runner.runWithMutex(0, "C"),
]);
```

Output:

```
C
B
A
A
B
C
```
