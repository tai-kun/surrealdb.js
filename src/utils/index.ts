export type * from "./src/diagnostics-channel";
export { default as channel } from "./src/diagnostics-channel";

export {
  BACKTICK,
  BRACKET_L,
  BRACKET_R,
  DOUBLE_QUOTE,
  escapeIdent,
  escapeKey,
  escapeRid,
  quoteStr,
  SINGLE_QUOTE,
} from "./src/escape";

export type * from "./src/is-array-buffer";
export { default as isArrayBuffer } from "./src/is-array-buffer";

export type * from "./src/get-timeout-signal";
export { default as getTimeoutSignal } from "./src/get-timeout-signal";

export type * from "./src/is-browser";
export { default as isBrowser } from "./src/is-browser";

export type * from "./src/is-live-result";
export { default as isLiveResult } from "./src/is-live-result";

export type * from "./src/is-rpc-response";
export { default as isRpcResponse } from "./src/is-rpc-response";

export type * from "./src/is-safe-number";
export { default as isSafeNumber } from "./src/is-safe-number";

export type * from "./src/make-abort-api";
export { default as makeAbortApi } from "./src/make-abort-api";

export type * from "./src/mutex";
export { default as mutex } from "./src/mutex";

export type * from "./src/serial";
export { default as Serial } from "./src/serial";

export type * from "./src/stateful-promise";
export { default as StatefulPromise } from "./src/stateful-promise";

export type * from "./src/task-emitter";
export { default as TaskEmitter } from "./src/task-emitter";

export type * from "./src/task-queue";
export { default as TaskQueue } from "./src/task-queue";

export type * from "./src/throw-if-aborted";
export { default as throwIfAborted } from "./src/throw-if-aborted";

export type * from "./src/to-surql";
export { default as toSurql } from "./src/to-surql";

export type * from "./src/types";

export type * from "./src/utf8";
export { default as utf8 } from "./src/utf8";
