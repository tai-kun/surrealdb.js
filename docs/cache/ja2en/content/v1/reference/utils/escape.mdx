---
title: escape
slug: en/v1/reference/utils/escape
---

## quoteStr

`quoteStr` is a function that encloses a string with quotation marks. For convenience, it encloses even an empty string with quotation marks.

### Import

```ts
import { quoteStr } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function quoteStr(str: string): string;
```

Encloses the string with quotation marks (`"` or `'`).

#### Arguments

##### `str`

The string to be enclosed in quotation marks.

#### Return Value

Returns the string enclosed in quotation marks. The quotation mark for enclosure changes depending on whether the given string contains a single quotation mark (`'`) or not:

- When the given string is empty
  - Returns `''`.
- When the given string does **not** contain a single quotation mark
  - Encloses it with a single quotation mark.
- When the given string does **contain** a single quotation mark
  - Encloses it with a double quotation mark.
- When the given string contains both a single and a double quotation mark
  - Encloses it with a double quotation mark.
  - The double quotation mark is escaped with a backslash (`\`).

#### Example

The following example encloses a string that **does not** contain a single quotation mark with a single quotation mark:

```ts
import { quoteStr } from "@tai-kun/surrealdb/utils";

console.log(quoteStr(``));
console.log(quoteStr(`a`));
console.log(quoteStr(`𩸽`));
console.log(quoteStr(`👍🏽`));
console.log(quoteStr(`パ`));
```

Output:

```js
''
'a'
'𩸽'
'👍🏽'
'パ'
```

The following example encloses a string that **does** contain a single quotation mark with a double quotation mark:

```ts
import { quoteStr } from "@tai-kun/surrealdb/utils";

console.log(quoteStr(`cat's`));
console.log(quoteStr(`𩸽's`));
console.log(quoteStr(`'👍🏽`));
console.log(quoteStr(`'パ`));
```

Output:

```js
"cat's"
"𩸽's"
"'👍🏽"
"'パ"
```

The following example encloses a string containing both a single and a double quotation mark with a double quotation mark and escapes if necessary:

```ts
import { quoteStr } from "@tai-kun/surrealdb/utils";

console.log(quoteStr(`cat's "toy"`));
console.log(quoteStr(`'\\"\\`));
console.log(quoteStr(`𩸽's "feed"`));
console.log(quoteStr(`'👍🏽"`));
console.log(quoteStr(`'パ"`));
```

Output:

```js
"cat's \"toy\""
"'\\\"\\"
"𩸽's \"feed\""
"'👍🏽\""
"'パ\""
```

## escapeKey

`escapeKey` is similar to [`quoteStr`](/en/reference/utils/quote-str/) but is a function used for object keys. For convenience, it encloses even an empty string with quotation marks.

### Import

```ts
import { escapeKey } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function escapeKey(key: string): string;
```

#### Arguments

##### `key`

The object's key.

#### Return Value

Returns the escaped string. It encloses in double quotation marks if necessary:

- When the given string is empty
  - Returns `""`.
- When the given string is composed only of alphanumeric characters and underscores
  - Returns it as is.
- When the given string includes characters other than alphanumeric characters and underscores
  - Encloses it in double quotation marks.
- When the given string contains a backslash (`\`)
  - Escapes it.
- When the given string contains a double quotation mark
  - Escapes it.

#### Example

The following example gets the string as is because the string is composed only of alphanumeric characters and underscores:

```ts
import { escapeKey } from "@tai-kun/surrealdb/utils";

console.log(escapeKey(`123`));
console.log(escapeKey(`foo_bar`));
```

Output:

```js
123
foo_bar
```

The following example encloses the string with double quotation marks and escapes if necessary:

```ts
import { escapeKey } from "@tai-kun/surrealdb/utils";

console.log(escapeKey(``));
console.log(escapeKey(`foo-bar`));
console.log(escapeKey(`'\\"\\`));
console.log(escapeKey(`𩸽's "feed"`));
console.log(escapeKey(`'👍🏽"`));
console.log(escapeKey(`'パ"`));
```

Output:

```js
""
"foo-bar"
"'\\"\"
"𩸽's \"feed\""
"'👍🏽\""
"'パ\""
```

## escapeRid

`escapeRid` is similar to [`quoteStr`](/en/reference/utils/quote-str/) but is a function used for record IDs. For convenience, it encloses even an empty string with quotation marks.

### Import

```ts
import { escapeRid } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function escapeRid(rid: string): string;
```

#### Arguments

##### `rid`

The ID part of the record ID.

#### Return Value

Returns the escaped string. It encloses in brackets (`⟨⟩`) if necessary:

- When the given string is empty
  - Returns `⟨⟩`.
- When the given string does not start with a digit and is composed only of alphanumeric characters and underscores
  - Returns it as is.
- When the given string contains a bracket (`⟩`)
  - Escapes it.
- When the given string starts with a digit or contains characters other than alphanumeric characters and underscores
  - Encloses it in brackets (`⟨⟩`).

#### Example

The following example gets the string as is because the string does not start with a digit and is composed only of alphanumeric characters and underscores

```ts
import { escapeRid } from "@tai-kun/surrealdb/utils";

console.log(escapeRid(`foo_123`));
```

Output:

```js
foo_123
```

The following example encloses the string in brackets (`⟨⟩`) and escapes if necessary:

```ts
import { escapeRid } from "@tai-kun/surrealdb/utils";

console.log(escapeRid(``));
console.log(escapeRid(`123`));
console.log(escapeRid(`foo-bar`));
console.log(escapeRid(`𩸽's ⟨feed⟩`));
console.log(escapeRid(`'👍🏽⟩`));
console.log(escapeRid(`'パ⟩`));
```

Output:

```js
⟨⟩
⟨123⟩
⟨foo-bar⟩
⟨𩸽's ⟨feed\⟩⟩
⟨'👍🏽\⟩⟩
⟨'パ\⟩⟩
```

## escapeIdent

`escapeIdent` is similar to [`quoteStr`](/en/reference/utils/quote-str/) but is a function used for the names of namespaces, databases, tables, and fields. For convenience, it encloses even an empty string with quotation marks.

### Import

```ts
import { escapeIdent } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function escapeIdent(ident: string): string;
```

#### Arguments

##### `ident`

The identifier.

#### Return Value

Returns the escaped string. It encloses in backticks (<code>`</code>) if necessary:

- When the given string is empty
  - Returns <code>``</code>.  {/* TODO(tai-kun): " になる問題解決する */}
- When the given string does not start with a digit and is composed only of alphanumeric characters and underscores
  - Returns it as is.
- When the given string contains a backtick (<code>`</code>)
  - Escapes it.
- When the given string starts with a digit or contains characters other than alphanumeric characters and underscores
  - Encloses it in backticks (<code>`</code>).

#### Example

The following example gets the string as is because the string does not start with a digit and is composed only of alphanumeric characters and underscores

```ts
import { escapeIdent } from "@tai-kun/surrealdb/utils";

console.log(escapeIdent(`foo_123`));
```

Output:

```js
foo_123
```

The following example encloses the string in backticks (<code>`</code>) and escapes if necessary:

```ts
import { escapeIdent } from "@tai-kun/surrealdb/utils";

console.log(escapeIdent(""));
console.log(escapeIdent("123"));
console.log(escapeIdent("foo-bar"));
console.log(escapeIdent("𩸽's `feed`"));
console.log(escapeIdent("'👍🏽`"));
console.log(escapeIdent("'パ`"));
```

Output:

```js
``
`123`
`foo-bar`
`𩸽's \`feed\``
`'👍🏽\``
`'パ\``
```

## Column

### String Length

The strings used in the previous examples contain characters that have different string lengths from the JavaScript perspective and the human perspective:

| Type | Display | `.length` | `Uint8Array` |
| --- | --- | --- | --- |
| ASCII | a | 1 | `[ 0x61 ]` |
| Surrogate Pair | 𩸽 | 2 | `[ 0xF0, 0xA9, 0xB8, 0xBD ]` |
| Variation Selector | 👍🏽 | 4 | `[ 0xF0, 0x9F, 0x91, 0x8D, 0xF0, 0x9F, 0x8F, 0xBD ]` |
| Combining Character | パ | 2 | `[ 0xE3, 0x83, 0x8F, 0xE3, 0x82, 0x9A ]` |

If the value of `.length` is counterintuitive, you need to be careful about operations on strings.
