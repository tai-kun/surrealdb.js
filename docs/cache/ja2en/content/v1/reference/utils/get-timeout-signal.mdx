---
title: getTimeoutSignal
slug: en/v1/reference/utils/get-timeout-signal
---

`getTimeoutSignal` is a function that returns an `AbortSignal` object that will timeout after the specified number of milliseconds. It is provided for environments where `AbortSignal.timeout()` is not implemented.

### Import

```ts
import { getTimeoutSignal } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function getTimeoutSignal(milliseconds: number): AbortSignal;
```

#### Arguments

##### `milliseconds`

Sets the number of milliseconds to wait before timing out.

#### Return value

Returns a new `AbortSignal` object.

#### Exceptions

Throws an error if the `AbortSignal.timeout` function is not present in the server-side runtime. This is because the cost of implementing a polyfill is high. Browsers implement it in a somewhat hacky way.

#### Example

The following example shows how to use `getTimeoutSignal` to timeout a `fetch` call after 5 seconds:

```ts
import { getTimeoutSignal } from "@tai-kun/surrealdb/utils";

const resp = await fetch("https://example.com/", {
  signal: getTimeoutSignal(5000),
});
```
