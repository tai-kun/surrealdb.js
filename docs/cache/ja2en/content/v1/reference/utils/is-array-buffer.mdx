---
title: isArrayBuffer
---

The `isArrayBuffer` function determines whether a given value is an `ArrayBuffer` object.

### Import

```ts
import { isArrayBuffer } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function isArrayBuffer(value: unknown): value is ArrayBuffer;
```

#### Arguments

##### `value`

The value expected to be an `ArrayBuffer` object.

#### Returns

Returns `true` if the `value` is an `ArrayBuffer` object, `false` otherwise.

### Explanation

In certain Firefox versions, converting a `fetch` response body to an `ArrayBuffer` object may result in an object that has the methods of an `ArrayBuffer` but fails the `instanceof` check. This function checks the tag name for a more reliable determination, similar to the behavior in lodash.
