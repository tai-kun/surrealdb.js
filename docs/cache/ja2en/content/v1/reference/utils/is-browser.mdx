---
title: isBrowser
slug: en/v1/reference/utils/is-browser
---

The `isBrowser` function determines whether the current JavaScript runtime is running in a browser environment.

### Import

```ts
import { isBrowser } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function isBrowser(): boolean;
```

#### Arguments

None.

#### Return value

Returns `true` if the runtime is in a browser environment, `false` otherwise.

#### Example

The following example checks if the current environment is a browser:

```ts
import { isBrowser } from "@tai-kun/surrealdb/utils";

if (isBrowser()) {
  document;
}
```

### Explanation

The function checks if the global variable `document` is defined. If it is not `undefined`, it is assumed to be running in a browser environment. For example, Deno adds a `window` object to the global scope, so checking for the existence of `window` would incorrectly identify a server runtime as a browser environment. Web frameworks like Remix call this [Document Guard](https://remix.run/docs/en/main/guides/constraints#document-guard).
