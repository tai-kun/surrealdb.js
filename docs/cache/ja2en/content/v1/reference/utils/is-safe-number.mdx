---
title: isSafeNumber
---

The `isSafeNumber` function checks if a given number is safe for arithmetic operations.

### Import

```ts
import { isSafeNumber } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function isSafeNumber(value: number): boolean;
```

#### Arguments

##### `value`

The number to be validated.

#### Return Value

Returns `false` if `value` is:

- `NaN`
- Infinity
- Outside the range of Â±2^53-1
