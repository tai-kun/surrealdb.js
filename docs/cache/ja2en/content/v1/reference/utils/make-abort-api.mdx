---
title: makeAbortApi
slug: en/v1/reference/utils/make-abort-api
---

`makeAbortApi` is a function that creates an abort signal and its associated abort function.

### Import

```ts
import { makeAbortApi } from "@tai-kun/surrealdb/utils";
```

### Usage

```ts
function makeAbortApi(signal?: AbortSignal): [
  signal: AbortSignal,
  abort: (reason?: unknown) => void,
];
```

#### Arguments

##### `signal`

The abort signal to associate with.

#### Return Value

Returns an array containing the following in order:

##### `signal`

The abort signal. It is not initially aborted.

##### `abort`

Dispatches an `"abort"` event to the abort signal. Optionally takes one argument `reason` which can be used to communicate the reason for the abort through the `"abort"` event.

#### Example

The following example creates an Abort API without any arguments:

```ts
import { makeAbortApi } from "@tai-kun/surrealdb/utils";

const [signal, abort] = makeAbortApi();
abort();
```

The example above is mostly equivalent to `new AbortController()`. The difference between `AbortController` and `makeAbortApi` is that `makeAbortApi` returns an array instead of an object.

The following example creates an Abort API with an associated abort signal:

```ts
import { makeAbortApi } from "@tai-kun/surrealdb/utils";

const controller = new AbortController();
controller.signal.addEventListener("abort", function() {
  console.log("controller.signal", this.reason);
});

const [signal] = makeAbortApi(controller.signal);
signal.addEventListener("abort", function() {
  console.log("signal", this.reason);
});

controller.abort("test");
```

Output:

```js
controller.signal test
signal test
```

The abort event propagates to the associated abort signal even when aborted from the Abort API:

```ts
import { makeAbortApi } from "@tai-kun/surrealdb/utils";

const controller = new AbortController();
controller.signal.addEventListener("abort", function() {
  console.log("controller.signal", this.reason);
});

const [signal, abort] = makeAbortApi(controller.signal);
signal.addEventListener("abort", function() {
  console.log("signal", this.reason);
});

abort("test");
```

Output:

```js
controller.signal test
signal test
```
