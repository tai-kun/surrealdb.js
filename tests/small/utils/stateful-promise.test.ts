import { StatefulPromise } from "@tai-kun/surrealdb/utils";
import { describe, expect, test, vi } from "vitest";

test("`resolve` で解決できる", async () => {
  const { promise, resolve } = StatefulPromise.withResolvers<string>();

  expect(promise.state).toBe("pending");

  resolve("test");

  expect<string>(await promise).toBe("test");
  expect(promise.state).toBe("fulfilled");
});

test("`reject` で拒否できる", async () => {
  const { promise, reject } = StatefulPromise.withResolvers();

  expect(promise.state).toBe("pending");

  reject(new Error("test"));

  await expect(async () => await promise)
    .rejects
    .toThrowErrorMatchingInlineSnapshot("[Error: test]");
  expect(promise.state).toBe("rejected");
});

test("解決後は常に同じ値を返す", async () => {
  const promise = new StatefulPromise(resolve => resolve({}));
  const [a, b] = await Promise.all([promise, promise]);

  expect(a).toBe(b);
});

test("onFulfilled 内で投げられた例外は onRejected で補足されない", async () => {
  const fn = async () => {
    await StatefulPromise.resolve().then(
      () => {
        throw 0;
      },
      () => {
        // unreachable
        // 呼び出されればエラーを握りつぶす。
      },
    );
  };

  await expect(fn).rejects.toBe(0);
});

test("Promise で解決できる", async () => {
  const { promise, resolve } = StatefulPromise.withResolvers<string>();

  expect(promise.state).toBe("pending");

  resolve(Promise.resolve("test"));
  const handleFulfilled = vi.fn();
  promise.then(handleFulfilled);
  await vi.waitUntil(() => handleFulfilled.mock.calls.length > 0);

  expect(handleFulfilled.mock.calls).toStrictEqual([
    ["test"],
  ]);
});

test.fails("(仕様確認) .try 無しで同期的に例外を投げると reject 判定にならない", async () => {
  const fn = () => {
    throw new Error("test");
  };
  const tryFn = fn;
  const fnTest = async () => {
    await expect(tryFn)
      .rejects
      .toThrowErrorMatchingInlineSnapshot("[Error: test]");
  };

  await fnTest();
});

test(".try 有りで同期的に例外を投げると reject 判定になる", async () => {
  const fn = () => {
    throw new Error("test");
  };
  const tryFn = () => StatefulPromise.try(fn);
  const fnTest = async () => {
    await expect(tryFn)
      .rejects
      .toThrowErrorMatchingInlineSnapshot("[Error: test]");
  };

  await fnTest();
});

test("拒否された StatefulPromise の理由を収集する", async () => {
  const errors = await StatefulPromise.allRejected([
    StatefulPromise.resolve(1),
    StatefulPromise.reject("reason"),
    StatefulPromise.reject(3150),
    undefined,
    1129,
  ]);

  expect(errors).toStrictEqual([
    "reason",
    3150,
  ]);
});

describe("新しいインスタンスは継承先のクラスで構築される", async () => {
  class MyStatefulPromise<T> extends StatefulPromise<T> {}

  test("constructor", async () => {
    const promise = new MyStatefulPromise<string>(resolve => resolve(""))
      .then(() => 0);

    expect(promise).toBeInstanceOf(MyStatefulPromise);
    expect<number>(await promise).toBe(0);
  });

  test(".resolve", async () => {
    const promise = MyStatefulPromise.resolve(0);

    expect(promise).toBeInstanceOf(MyStatefulPromise);
  });

  test(".reject", async () => {
    const promise = MyStatefulPromise.reject(0);

    expect(promise).toBeInstanceOf(MyStatefulPromise);
  });

  test(".withResolvers", async () => {
    const { promise, resolve } = MyStatefulPromise.withResolvers<void>();
    resolve();

    expect(promise).toBeInstanceOf(MyStatefulPromise);
  });

  test(".try", async () => {
    const promise = MyStatefulPromise.try(x => x, 0);

    expect(promise).toBeInstanceOf(MyStatefulPromise);
    expect<number>(await promise).toBe(0);
  });

  test(".allRejected", async () => {
    const promise = MyStatefulPromise.allRejected([]);

    expect(promise).toBeInstanceOf(MyStatefulPromise);
  });
});

describe("ドキュメントの例", () => {
  test("解決", async () => {
    const spy = vi.spyOn(console, "log").mockImplementation(() => {});

    try {
      // Example code
      {
        const promise = new StatefulPromise(resolve => {
          setTimeout(() => resolve("test"), 0);
        });

        const result = await promise;

        console.log(result); // "test"
      }

      // Test
      {
        expect(spy.mock.calls).toStrictEqual([
          ["test"],
        ]);
      }
    } finally {
      spy.mockClear();
    }
  });

  test("拒否", async () => {
    const spy = vi.spyOn(console, "log").mockImplementation(() => {});

    try {
      // Example code
      {
        const promise = new StatefulPromise((_, reject) => {
          setTimeout(() => reject("test"), 0);
        });

        while (promise.state === "pending") {
          await new Promise(r => setTimeout(r, 50));
        }

        try {
          await promise;
        } catch (e) {
          console.log(e); // "test"
        }
      }

      // Test
      {
        expect(spy.mock.calls).toStrictEqual([
          ["test"],
        ]);
      }
    } finally {
      // Clean up

      spy.mockClear();
    }
  });

  test(".try", async () => {
    const spy = vi.spyOn(console, "log").mockImplementation(() => {});

    try {
      // Example code
      {
        const promise = StatefulPromise.try(() => {
          throw "test";
        });

        await promise.then(null, e => {
          console.log(e); // "test"
        });
      }

      // Test
      {
        expect(spy.mock.calls).toStrictEqual([
          ["test"],
        ]);
      }
    } finally {
      // Clean up

      spy.mockClear();
    }
  });
});
