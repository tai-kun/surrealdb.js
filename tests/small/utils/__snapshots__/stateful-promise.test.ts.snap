// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`.try 有りで同期的に例外を投げると reject 判定になる 1`] = `
{
  "message": "test",
  "name": "Error",
}
`;

exports[`\`reject\` で拒否できる 1`] = `
{
  "message": "test",
  "name": "Error",
}
`;
